File: internal/detect/emit.go
package detect

import (
	"path/filepath"
	"sort"
	"strconv"

	"gopkg.in/yaml.v3"
)

// RepoDetection is everything detected for one cloned repo: its runnable apps
// (from DetectRepo) and its backing services (shared ComposeServices from
// ParseCompose).
type RepoDetection struct {
	Name   string
	Alias  string // optional; defaults to Name
	Apps   []StackFacts
	Shared []ComposeService
}

// Emit renders a draft pom.yml from detection results. It emits the structure
// (repos → services with run commands, setup, shared_services) — a starting point
// the onboarding agent refines (env wiring is left to Phase 1). The output is
// valid config: it loads and validates cleanly.
func Emit(session string, repos []RepoDetection) (string, error) {
	cfg := emitConfig{Session: session, DefaultBranch: "main"}
	sharedSeen := map[string]*emitShared{}

	for _, rd := range repos {
		al := rd.Alias
		if al == "" {
			al = rd.Name
		}
		er := &emitRepo{Alias: al}
		used := map[string]bool{}
		var setup []string
		setupSeen := map[string]bool{}

		for _, app := range rd.Apps {
			if app.Install != "" && !setupSeen[app.Install] {
				setupSeen[app.Install] = true
				setup = append(setup, app.Install)
			}
			for _, s := range app.Setup {
				if !setupSeen[s] {
					setupSeen[s] = true
					setup = append(setup, s)
				}
			}
			for _, run := range app.Run {
				name := serviceName(app, run, used)
				used[name] = true
				svc := &emitService{Cmd: run.Cmd}
				if app.Dir != "" {
					svc.Dir = app.Dir
				}
				if run.Kind == "server" && app.Port > 0 {
					yes := true
					svc.Port = &yes
				}
				if er.Services == nil {
					er.Services = map[string]*emitService{}
				}
				er.Services[name] = svc
			}
		}
		er.Setup = setup

		for _, sh := range rd.Shared {
			if sh.Kind != KindShared || sh.Type == "custom" || sh.Type == "" {
				continue
			}
			er.SharedServices = appendUniq(er.SharedServices, sh.Type)
			if _, ok := sharedSeen[sh.Type]; !ok {
				sharedSeen[sh.Type] = &emitShared{Type: sh.Type}
			}
		}
		sort.Strings(er.SharedServices)

		if cfg.Repos == nil {
			cfg.Repos = map[string]*emitRepo{}
		}
		cfg.Repos[rd.Name] = er
	}

	if len(sharedSeen) > 0 {
		cfg.SharedServices = sharedSeen
	}

	out, err := yaml.Marshal(cfg)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

type emitConfig struct {
	Session        string                 `yaml:"session,omitempty"`
	DefaultBranch  string                 `yaml:"default_branch,omitempty"`
	Repos          map[string]*emitRepo   `yaml:"repos,omitempty"`
	SharedServices map[string]*emitShared `yaml:"shared_services,omitempty"`
}

type emitRepo struct {
	Alias          string                  `yaml:"alias,omitempty"`
	Services       map[string]*emitService `yaml:"services,omitempty"`
	Setup          []string                `yaml:"setup,omitempty"`
	SharedServices []string                `yaml:"shared_services,omitempty"`
}

type emitService struct {
	Cmd  string `yaml:"cmd"`
	Dir  string `yaml:"dir,omitempty"`
	Port *bool  `yaml:"port,omitempty"`
}

type emitShared struct {
	Type string `yaml:"type,omitempty"`
}

// serviceName derives a stable, unique service name from an app + run.
func serviceName(app StackFacts, run ResolvedRun, used map[string]bool) string {
	base := app.Framework
	if app.Dir != "" {
		base = filepath.Base(app.Dir)
	}
	if base == "" {
		base = app.Language
	}
	if base == "" {
		base = "app"
	}
	if run.Kind == "worker" {
		base += "-worker"
	}
	name := base
	for i := 2; used[name]; i++ {
		name = base + "-" + strconv.Itoa(i)
	}
	return name
}

func appendUniq(s []string, v string) []string {
	for _, x := range s {
		if x == v {
			return s
		}
	}
	return append(s, v)
}

File: rust/crates/pom_detect/src/emit.rs
use std::collections::BTreeMap;
use std::path::Path;

use crate::{ComposeService, ResolvedRun, ServiceKind, StackFacts};

/// One cloned repo's findings: its apps and its backing services.
#[derive(Clone, Debug, Default)]
pub struct RepoDetection {
    pub name: String,
    /// Defaults to the name.
    pub alias: String,
    pub apps: Vec<StackFacts>,
    pub shared: Vec<ComposeService>,
}

fn service_name(app: &StackFacts, run: &ResolvedRun, used: &[String]) -> String {
    let mut base = if app.dir.is_empty() {
        app.framework.clone()
    } else {
        Path::new(&app.dir)
            .file_name()
            .map(|name| name.to_string_lossy().into_owned())
            .unwrap_or_default()
    };
    if base.is_empty() {
        base = app.language.clone();
    }
    if base.is_empty() {
        base = "app".into();
    }
    if run.kind == "worker" {
        base.push_str("-worker");
    }
    let mut name = base.clone();
    let mut index = 2;
    while used.contains(&name) {
        name = format!("{base}-{index}");
        index += 1;
    }
    name
}

/// A YAML scalar, quoted unless it is plainly safe.
fn scalar(text: &str) -> String {
    let plain = !text.is_empty()
        && text
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || " _-./@".contains(c))
        && !text.starts_with([' ', '-', '@', '.'])
        && !text.ends_with(' ')
        && !matches!(
            text.to_ascii_lowercase().as_str(),
            "true" | "false" | "yes" | "no" | "on" | "off" | "null" | "~"
        )
        && text.parse::<f64>().is_err();
    if plain {
        text.to_string()
    } else {
        format!("\"{}\"", text.replace('\\', "\\\\").replace('"', "\\\""))
    }
}

/// A draft `pom.yml` from what was detected: repos with their services, setup and shared services. It loads
/// and validates; env wiring is left to the onboarding agent.
pub fn emit(session: &str, repos: &[RepoDetection]) -> String {
    let mut out = format!("session: {}\ndefault_branch: main\n", scalar(session));
    let mut shared_types: BTreeMap<String, ()> = BTreeMap::new();
    let mut sorted: Vec<&RepoDetection> = repos.iter().collect();
    sorted.sort_by(|a, b| a.name.cmp(&b.name));
    if !sorted.is_empty() {
        out.push_str("repos:\n");
    }
    for repo in sorted {
        out.push_str(&format!("  {}:\n", scalar(&repo.name)));
        let alias = if repo.alias.is_empty() {
            &repo.name
        } else {
            &repo.alias
        };
        out.push_str(&format!("    alias: {}\n", scalar(alias)));
        let mut used: Vec<String> = Vec::new();
        let mut setup: Vec<String> = Vec::new();
        let mut services: BTreeMap<String, (String, String, bool)> = BTreeMap::new();
        for app in &repo.apps {
            for step in std::iter::once(&app.install).chain(app.setup.iter()) {
                if !step.is_empty() && !setup.contains(step) {
                    setup.push(step.clone());
                }
            }
            for run in &app.run {
                let name = service_name(app, run, &used);
                used.push(name.clone());
                services.insert(
                    name,
                    (
                        run.cmd.clone(),
                        app.dir.clone(),
                        run.kind == "server" && app.port > 0,
                    ),
                );
            }
        }
        if !services.is_empty() {
            out.push_str("    services:\n");
            for (name, (cmd, dir, port)) in &services {
                out.push_str(&format!(
                    "      {}:\n        cmd: {}\n",
                    scalar(name),
                    scalar(cmd)
                ));
                if !dir.is_empty() {
                    out.push_str(&format!("        dir: {}\n", scalar(dir)));
                }
                if *port {
                    out.push_str("        port: true\n");
                }
            }
        }
        if !setup.is_empty() {
            out.push_str("    setup:\n");
            for step in &setup {
                out.push_str(&format!("      - {}\n", scalar(step)));
            }
        }
        let mut shared: Vec<String> = repo
            .shared
            .iter()
            .filter(|service| {
                service.kind == ServiceKind::Shared
                    && !service.kind_name.is_empty()
                    && service.kind_name != "custom"
            })
            .map(|service| service.kind_name.clone())
            .collect();
        shared.sort();
        shared.dedup();
        if !shared.is_empty() {
            out.push_str("    shared_services:\n");
            for kind in &shared {
                out.push_str(&format!("      - {}\n", scalar(kind)));
                shared_types.insert(kind.clone(), ());
            }
        }
    }
    if !shared_types.is_empty() {
        out.push_str("shared_services:\n");
        for kind in shared_types.keys() {
            out.push_str(&format!(
                "  {}:\n    type: {}\n",
                scalar(kind),
                scalar(kind)
            ));
        }
    }
    out
}
